torch/testing/_internal/common_distributed.py

```diff
@@ -167,40 +168,37 @@ class MultiProcessTestCase(TestCase):
     def world_size(self):
         return 4
 
-    @staticmethod
-    def join_or_run(fn):
+    def join_or_run(self, fn):
         @wraps(fn)
         def wrapper(self):
             if self.rank == self.MAIN_PROCESS_RANK:
                 self._join_processes(fn)
             else:
                 try:
-                    fn(self)
+                    fn()
                 except Exception as e:
                     logging.error('Caught exception: \n{}exiting process with exit code: {}'
                                   .format(traceback.format_exc(), MultiProcessTestCase.TEST_ERROR_EXIT_CODE))
                     sys.exit(MultiProcessTestCase.TEST_ERROR_EXIT_CODE)
-        return wrapper
+        return types.MethodType(wrapper, self)
 
     # The main process spawns N subprocesses that run the test.
-    # This function patches overwrites every test function to either
+    # Constructor patches current instance test method to
     # assume the role of the main process and join its subprocesses,
     # or run the underlying test function.
-    @classmethod
-    def setUpClass(cls):
-        for attr in dir(cls):
-            if attr.startswith('test'):
-                fn = getattr(cls, attr)
-                setattr(cls, attr, cls.join_or_run(fn))
+    def __init__(self, method_name='runTest'):
```

Comment:
Curious, where does the default value `runTest` come from? And I assume this will be replaced by the real test function name for each individual test run? I am not familiar with `TestCase` implementation, would I be correct if I assume it is guaranteed to create a new `MultiProcessTestCase` instance (and hence call this ctor with the proper function) for each test function run regardless of how we launch tests?

Reply:
Yes, from https://docs.python.org/3/library/unittest.html#unittest.TestCase :

> Each instance of TestCase will run a single base method: the method named methodName. >In most uses of TestCase, you will neither change the methodName nor reimplement the >default runTest() method.